logo.js

```diff
@@ -841,19 +911,78 @@ function LogoInterpreter(turtle, stream, savehook)
       var atoms = parse(string);
 
       // And execute it!
-      return self.execute(atoms, options);
+      var promise = self.execute(atoms, options).then(null, function (err) {
+        if (err && err.special == "bye") {
+          return undefined;
+        }
+        if (self.turtle) {
+          self.turtle.end();
+        }
+        throw err;
+      });
+      this._setPendingPromise(promise);
+      self._pendingPromise = promise;
+      return promise;
     } catch (e) {
+      // FIXME: doesn't work with promises
       if (e instanceof Bye) {
         // clean exit
-        return undefined;
+        return Promise.resolve(undefined);
       } else {
-        throw e;
+        return Promise.reject(e);
       }
     } finally {
       if (self.turtle) { self.turtle.end(); }
     }
   };
 
+  //----------------------------------------------------------------------
+  // Helpers to ensure calls to run() are serialized:
+  //----------------------------------------------------------------------
+
+  // When there's an outstanding call to .run() (whose promise hasn't resolved
```

Comment:
Just to confirm - the idea here is that if the promise returned by run() hasn't settled (e.g. it's waiting on a newfangled async `READWORD` prompt, or a long-running turtle operation, etc), and another run() comes in, we enqueue it then run it later?

Can I challenge that use case and say, for simplicity, that run() throws/rejects at that point, and its up to the caller to prevent that - e.g. the UI is disabled until the first `run` settles? In the UI we could turn Run into Stop which would interrupt execution in the case of an infinite loop. (Not handled today, I don't think)